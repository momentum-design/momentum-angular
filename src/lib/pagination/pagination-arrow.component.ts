import { Component, OnInit, Input, OnDestroy } from '@angular/core';
import { PaginationService } from './pagination.service';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

@Component({
  selector: 'md-pagination-arrow',
  template: `
    <a #btn *ngIf="isEnable;else disableTemplate"
      [ngClass]='ngClassA'
      [href]='myHref'
      [attr.aria-label]='name'
      (click)='handleClick($event)'
    ><span class="show-for-sr">page</span></a>
    <ng-template #disableTemplate>
      <span class="show-for-sr">page</span>
    </ng-template>
  `,
  styleUrls: ['pagination.scss'],
  host: {
    'class': 'pagination-previous pagination_li md-icon icon',
    '[class.pagination_li_last]': '!isPrevious',
    '[class.pagination_disable]': '!isEnable',
    '[class.icon-arrow-left-optical_20]': 'isPrevious && !isEnable',
    '[class.icon-arrow-right-optical_20]': '!isPrevious && !isEnable'
  }
})
export class PaginationArrowComponent implements OnInit, OnDestroy {
  /** @option set isPrevious | '' */
  @Input() isPrevious: boolean = true;
  /** @option set ifPreventDefault | '' */
  @Input() ifPreventDefault: boolean = false;

  public isEnable = false;
  public name = 'Previous page';
  public myHref: any = '';
  public ngClassA = {
    'pagination_li_a': true,
    'icon--b1': true,
    'md-icon': true,
    'icon': true,
    'pagination_li_a_focus': false,
    'icon-arrow-left_20': true,
    'icon-arrow-right_20': false
  };

  private unsubscribe$ = new Subject();

  constructor(private paginationService: PaginationService) {
    this.paginationService.current$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe(() => {
      this.build();
    });
    this.paginationService.total$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe(() => {
      this.build();
    });
    this.paginationService.focused$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      if (this.isPrevious) {
        this.ngClassA.pagination_li_a_focus = v === 'previous';
      } else {
        this.ngClassA.pagination_li_a_focus = v === 'next';
      }
    });
  }

  ngOnInit() {
    this.build();
  }

  public handleClick = (event) => {
    if (this.isPrevious) {
      this.paginationService.focus('previous');
      this.paginationService.selectPrevious();
    } else {
      this.paginationService.focus('next');
      this.paginationService.selectNext();
    }
    event.stopPropagation();
    if (this.ifPreventDefault) {
      event.preventDefault();
    }
  }

  public build = () => {
    const s = this.paginationService;
    if (this.isPrevious) {
      this.isEnable = !s.isCurrentFirst();
      this.name = 'Previous page';
      this.ngClassA['icon-arrow-left_20'] = true;
      this.ngClassA['icon-arrow-right_20'] = false;
      this.myHref = s.getHref(s.currentPage - 1);
    } else {
      this.isEnable = !s.isCurrentLast();
      this.name = 'Next page';
      this.ngClassA['icon-arrow-left_20'] = false;
      this.ngClassA['icon-arrow-right_20'] = true;
      this.myHref = s.getHref(s.currentPage + 1);
    }
  }

  ngOnDestroy() {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }
}
