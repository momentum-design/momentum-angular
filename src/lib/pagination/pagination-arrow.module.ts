import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PaginationArrowComponent } from './pagination-arrow.component';

@NgModule({
  declarations: [PaginationArrowComponent],
  imports: [
    CommonModule
  ],
  exports: [PaginationArrowComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class PaginationArrowModule { }
