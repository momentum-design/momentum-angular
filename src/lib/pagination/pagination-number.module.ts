import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PaginationNumberComponent } from './pagination-number.component';

@NgModule({
  declarations: [PaginationNumberComponent],
  imports: [
    CommonModule
  ],
  exports: [PaginationNumberComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class PaginationNumberModule { }
