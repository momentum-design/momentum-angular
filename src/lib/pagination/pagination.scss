.ellipsis {
  width: 11px;
  height: 16px;
  color: rgb(102, 102, 102);
  font-size: 12px;
  font-family: CiscoSans;
  text-align: center;
  letter-spacing: 0px;
}

.pagination_li{
  display: block;
  float: left;
  font-size: 12px;
  margin:0 20px 0 0;
  height: 20px;
  line-height: 20px;
  min-width: 20px;
  border-radius: 10px !important;
  text-align: center !important;
}

.pagination_li_last{
  margin-right:0px;
}

.pagination_li_a{
  transition: background-color 300ms ease-out;
  display: block;
  font-size: 12px;
  color: rgb(5, 161, 191);
  border-radius: 10px !important;
  text-align: center !important;
  padding: 0 4px;
  line-height: 20px;
}

.pagination_li_a:focus, .pagination_li_a:hover, .pagination_current{
  background: rgb(7, 193, 228);
  color: white;
  padding: 0 4px;
}

.pagination_disable{
  color: rgba(0,0,0,0.4);
  padding: 0 4px;
}

.pagination_li_a_focus, .pagination_li_focus{
  border-color: transparent;
  outline: none;
  box-shadow: 0 0 4px 2px rgba(0, 160, 209, 0.75);
}
