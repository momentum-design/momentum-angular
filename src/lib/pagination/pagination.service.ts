import { Injectable, OnDestroy } from '@angular/core';
import { Subject } from 'rxjs';
import { DomSanitizer, SafeUrl } from '@angular/platform-browser';
import { takeUntil } from 'rxjs/operators';

@Injectable()
export class PaginationService implements OnDestroy {
  public href: string = '';
  public hrefReplaceReg: string | RegExp = '$page$';

  public currentPage: number;
  public focusPage: any;
  public totalPage: number;

  private current = new Subject<number>();
  private focused = new Subject<any>();
  private total = new Subject<number>();
  private unsubscribe$ = new Subject();

  current$ = this.current.asObservable();
  focused$ = this.focused.asObservable();
  total$ = this.total.asObservable();

  constructor(private sanitizer: DomSanitizer) {
    this.current$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      this.currentPage = v;
    });
    this.focused$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      this.focusPage = v;
    });
    this.total$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      this.totalPage = v;
    });
  }

  initConfig(href, hrefReplaceReg, totalPage) {
    this.href = href;
    this.hrefReplaceReg = hrefReplaceReg;
    this.totalPage = totalPage;
  }

  isClickable(index) {
    return index !== this.currentPage && index > 0 && index <= this.totalPage;
  }

  isFirstOrLastNow() {
    return this.currentPage === 0 || this.currentPage === this.totalPage;
  }

  isCurrent(index) {
    return this.currentPage === index;
  }

  isCurrentFirst() {
    return this.currentPage === 1;
  }

  isCurrentLast() {
    return this.currentPage === this.totalPage;
  }

  select(index) {
    if (this.isClickable(index)) {
      this.current.next(index);
    }
  }

  selectNext() {
    this.select(this.currentPage + 1);
  }

  selectPrevious() {
    this.select(this.currentPage - 1);
  }

  focus(index) {
    if (index !== this.focusPage) {
      this.focused.next(index);
    }
  }

  updateTotal(total) {
    this.total.next(total);
    if (this.totalPage < this.currentPage) {
      this.current.next(total);
    }
  }

  public getHref = (index): String | SafeUrl => {
    if (this.href !== '' && this.href) {
      return this.href.replace(this.hrefReplaceReg, index);
    } else {
      return this.sanitizer.bypassSecurityTrustUrl('javascript:void(0)');
    }
  }

  public focusInList = (list, offset) => {
    let index = list.indexOf(this.focusPage || this.currentPage);
    const lastIndex = list.length - 1;
    if (lastIndex > -1) {
      if (index === -1) {
        index = 0;
      }
      if (offset > 0) {
        index = Math.min(lastIndex, index + offset);
      } else {
        index = Math.max(0, index + offset);
      }
      this.focus(list[index]);
    }
  }

  public selectfocus = () => {
    if (this.focusPage === 'previous') {
      this.selectPrevious();
    } else if (this.focusPage === 'next') {
      this.selectNext();
    } else {
      this.select(this.focusPage);
    }
  }

  ngOnDestroy(): void {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }
}
