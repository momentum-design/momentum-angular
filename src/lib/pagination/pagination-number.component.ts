import { Component, OnInit, Input, OnDestroy } from '@angular/core';
import { PaginationService } from './pagination.service';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

@Component({
  selector: 'md-pagination-number',
  template: `
    <a [ngClass]='ngClassA' *ngIf="!isCurrent;else other_content"
    [href]='myHref'
    [attr.aria-label]="'Page '+index"
    (click)='handleClick($event)'
    >{{ index }}</a>
    <ng-template #other_content>{{ index }}</ng-template>
  `,
  styleUrls: ['pagination.scss'],
  host: {
    'class': 'pagination_li',
    '[class.pagination_current]': 'isCurrent',
    '[class.pagination_li_focus]': 'isCurrent && isFocus'
  }
})
export class PaginationNumberComponent implements OnInit, OnDestroy {
  /** @option set index | '' */
  @Input() index: number;
  /** @option set ifPreventDefault | '' */
  @Input() ifPreventDefault: boolean = false;

  public myHref: any;
  public isCurrent = false;
  public isFocus = false;
  public ngClassA = {
    'pagination_li_a': true,
    'pagination_li_a_focus': false
  };
  private unsubscribe$ = new Subject();

  constructor(
    private paginationService: PaginationService
  ) {
    this.paginationService.current$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe(() => {
      this.build();
    });
    this.paginationService.focused$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      this.isFocus = v === this.index;
      this.ngClassA.pagination_li_a_focus = this.isFocus;
    });
  }

  ngOnInit() {
    this.build();
  }

  private build = () => {
    const s = this.paginationService;
    this.myHref = s.getHref(this.index);
    this.isCurrent = s.isCurrent(this.index);
  }

  public handleClick = (event) => {
    this.paginationService.focus(this.index);
    this.paginationService.select(this.index);
    event.stopPropagation();
    if (this.ifPreventDefault) {
      event.preventDefault();
    }
  }

  ngOnDestroy () {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }
}
