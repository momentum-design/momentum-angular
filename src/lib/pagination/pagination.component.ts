import { Component, OnInit, Input, Output, EventEmitter, SimpleChanges, OnChanges, OnDestroy } from '@angular/core';
import { PaginationService } from './pagination.service';
import { ENTER, LEFT_ARROW, RIGHT_ARROW } from '@angular/cdk/keycodes';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

@Component({
  selector: 'md-pagination',
  template: `
    <md-pagination-arrow
      [isPrevious]='true'
      [ifPreventDefault]='ifPreventDefault'
    ></md-pagination-arrow>
    <md-pagination-number *ngFor='let item of renderGroupFirst'
      [index]='item'
      [ifPreventDefault]='ifPreventDefault'
    ></md-pagination-number>
    <li class='pagination_li ellipsis' *ngIf='firstGroupEnd > 0 && firstGroupEnd < midGroupStart - 1' aria-hidden="true">...</li>
    <md-pagination-number *ngFor='let item of renderGroupMid'
      [index]='item'
      [ifPreventDefault]='ifPreventDefault'
    ></md-pagination-number>
    <li class='pagination_li ellipsis' *ngIf='lastGroupStart <= total && lastGroupStart > midGroupEnd + 1' aria-hidden="true">...</li>
    <md-pagination-number *ngFor='let item of renderGroupLast'
      [index]='item'
      [ifPreventDefault]='ifPreventDefault'
    ></md-pagination-number>
    <md-pagination-arrow
      [isPrevious]='false'
      [ifPreventDefault]='ifPreventDefault'
    ></md-pagination-arrow>
  `,
  styleUrls: ['pagination.scss'],
  host: {
    class: 'pagination',
    'style': `
      display: block;
      min-height: 1.5rem;
      list-style-position: outside;
      clear:both;
    `,
    'attr.role': 'navigation',
    'attr.aria-label': 'Pagination',
    'tabindex': '0',
    '(keydown)': 'handleKeydown($event)'
  },
  providers: [PaginationService]
})
export class PaginationComponent implements OnInit, OnChanges, OnDestroy {

  /** @option set the href format | '' */
  @Input() href: string = '';
  /** @option set the replace placeholder of href */
  @Input() hrefReplaceReg: string | RegExp = '$page$';
  /** @props set total | '' */
  @Input() total: number = 10;
  /** @props set current | '' */
  @Input() current: number = 1;
  /** @props set the first button group | '' */
  @Input() firstGroupNum: number = 2;
  /** @props set the mid button group | '' */
  @Input() midGroupNum: number = 3;
  /** @props set the last button group | '' */
  @Input() lastGroupNum: number = 2;
  /** @option Callback function invoked when user clicks buttons */
  @Output() whenClick = new EventEmitter();

  public renderGroupFirst = [];
  public renderGroupMid = [];
  public renderGroupLast = [];
  public renderGroup = [];

  public firstGroupEnd: number = -1;
  public midGroupStart: number = 1;
  public midGroupEnd: number = 1;
  public lastGroupStart: number = Infinity;

  public ifPreventDefault = false;

  private minCurrentForFirstGroup = 0;
  private maxCurrentForLastGroup = 0;

  private unsubscribe$ = new Subject();

  constructor(
    private paginationService: PaginationService
  ) { }

  ngOnInit() {
    const s = this.paginationService;
    s.initConfig(this.href, this.hrefReplaceReg, this.total);
    s.current$
    .pipe(takeUntil(this.unsubscribe$))
    .subscribe((v) => {
      this.build();
      this.whenClick.emit(v);
    });
    this.paginationService.select(this.current);
    this.build();
  }

  ngOnChanges(changes: SimpleChanges) {
    let needsBuild = true;
    if (changes['total']) {
      needsBuild = changes['total'].currentValue >= this.paginationService.currentPage;
      this.paginationService.updateTotal(changes['total'].currentValue);
    }

    if (changes['current']) { // Will trigger a build
      this.paginationService.select(changes['current'].currentValue);
    } else if (needsBuild) {
        this.build();
    }

    this.paginationService.focus(this.paginationService.currentPage);
  }

  public handleKeydown = (event) => {
    const s = this.paginationService;
    let ifStop = false;

    switch (event.keyCode) {
      case LEFT_ARROW: // <-
        s.focusInList(this.renderGroup, -1);
        ifStop = true;
        break;
      case RIGHT_ARROW: // ->
        s.focusInList(this.renderGroup, 1);
        ifStop = true;
        break;
      case ENTER: // Enter
        s.selectfocus();
        ifStop = true;
        break;
      default:
        break;
    }
    if (ifStop) {
      event.stopPropagation();
    }
  }

  public build = () => {
    const s = this.paginationService,
      current = s.currentPage,
      total = s.totalPage,
      preMidLen = Math.floor((this.midGroupNum - 1) / 2),
      nextMidLen = this.midGroupNum - preMidLen - 1;
    this.minCurrentForFirstGroup = this.firstGroupNum + 1 + preMidLen; // > this
    this.maxCurrentForLastGroup = total - this.lastGroupNum - 1 - nextMidLen + 1; // > this

    if (current <= this.minCurrentForFirstGroup && current >= this.maxCurrentForLastGroup) { // merge all
      this.firstGroupEnd = -1;
      this.lastGroupStart = Infinity;
      this.midGroupStart = 1;
      this.midGroupEnd = total;
    } else {
      if (current > this.minCurrentForFirstGroup) {
        this.firstGroupEnd = this.firstGroupNum;
        this.midGroupStart = current - preMidLen;
        this.midGroupEnd = Math.min(total, this.midGroupStart + this.midGroupNum - 1);
      } else { // merge start and middle
        this.firstGroupEnd = -1;
        this.midGroupStart = 1;
        this.midGroupEnd = Math.min(total, Math.max(current + nextMidLen, this.minCurrentForFirstGroup));
      }

      if (current < this.maxCurrentForLastGroup) {
        this.lastGroupStart = total - this.lastGroupNum + 1;
      } else { // merge end and middle
        this.lastGroupStart = Infinity;
        this.midGroupStart = Math.max(1, Math.min(current - preMidLen, this.maxCurrentForLastGroup));
        this.midGroupEnd = total;
      }
    }

    // get template
    this.renderGroupFirst = [];
    this.renderGroupMid = [];
    this.renderGroupLast = [];
    this.renderGroup = [];
    let i = 1;

    if (current > 1) {
      this.renderGroup.push('previous');
    }

    for (i = 1; i <= this.firstGroupEnd; i++) {
      this.renderGroupFirst.push(i);
      this.renderGroup.push(i);
    }

    if (this.firstGroupEnd === this.midGroupStart) {
      this.renderGroupFirst.pop();
    }

    for (i = this.midGroupStart; i <= this.midGroupEnd; i++) {
      this.renderGroupMid.push(i);
      this.renderGroup.push(i);
    }

    if (this.midGroupEnd === this.lastGroupStart) {
      this.renderGroupMid.pop();
    }

    for (i = this.lastGroupStart; i <= total; i++) {
      this.renderGroupLast.push(i);
      this.renderGroup.push(i);
    }

    if (s.currentPage < total) {
      this.renderGroup.push('next');
    }
  }

  ngOnDestroy () {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }
}
