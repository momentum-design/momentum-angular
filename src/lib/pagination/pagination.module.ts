import { NgModule, CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PaginationComponent } from './pagination.component';
import { PaginationArrowModule } from './pagination-arrow.module';
import { PaginationNumberModule } from './pagination-number.module';

@NgModule({
  declarations: [PaginationComponent],
  imports: [
    CommonModule,
    PaginationArrowModule,
    PaginationNumberModule
  ],
  exports: [PaginationComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class PaginationModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: PaginationModule,
      providers: [],
    };
  }
}
