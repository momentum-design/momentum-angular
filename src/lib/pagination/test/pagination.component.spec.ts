import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { PaginationService } from '../pagination.service';
import { PaginationComponent } from '../pagination.component';
import { PaginationArrowModule } from '../pagination-arrow.module';
import { PaginationNumberModule } from '../pagination-number.module';
import { DebugElement, Component } from '@angular/core';
import { By } from '@angular/platform-browser';
import { PaginationNumberComponent } from '../pagination-number.component';
import { PaginationArrowComponent } from '../pagination-arrow.component';

@Component({
  template: `
    <md-pagination
      [total]='totalPages'
      [current]='currentPage'
      [firstGroupNum]='firstGroupNum'
      [midGroupNum]='midGroupNum'
      [lastGroupNum]='lastGroupNum'
    ></md-pagination>
  `
})
class PaginationWrapperComponent {
  totalPages: number;
  currentPage: number;
  firstGroupNum: number = 2;
  midGroupNum: number = 3;
  lastGroupNum: number = 2;
}

describe('PaginationComponent', () => {
  describe('creation and rendering', () => {
    let component: PaginationComponent;
    let fixture: ComponentFixture<PaginationComponent>;
    let testElements: Array<HTMLElement>;
    let paginationService: PaginationService;

    beforeEach(async(() => {
      TestBed.configureTestingModule({
        declarations: [ PaginationComponent ],
        imports: [PaginationArrowModule, PaginationNumberModule],
        providers: [ PaginationService ]
      })
      .compileComponents();
    }));

    beforeEach(() => {
      fixture = TestBed.createComponent(PaginationComponent);
      component = fixture.componentInstance;
      fixture.detectChanges();
    });

    it('should create and match snapshot', () => {
      expect(component).toBeTruthy();
      expect(fixture).toMatchSnapshot();
    });

    it('should create', () => {
      expect(component).toBeTruthy();
    });

    it('should run service Overlay', () => {
      component.ngOnInit();
      fixture.detectChanges();
      paginationService = TestBed.get(PaginationService);
      expect(paginationService).toBeTruthy();
    });

    it('should create 4 clickable elements', () => {
      component.total = 3;
      component.current = 2;
      component.ngOnInit();
      fixture.detectChanges();
      testElements = fixture.nativeElement.getElementsByTagName('a');

      // Clickable elements should be ( < 1 3 > )
      expect(testElements.length).toEqual(4);
    });

    it('should disable previous', () => {
      component.total = 3;
      component.ngOnInit();
      fixture.detectChanges();
      const prevArrow = fixture.debugElement.queryAll(By.directive(PaginationArrowComponent))[0];
      expect(prevArrow.componentInstance.isEnable).toBe(false);
    });

    it('should correctly render 5 buttons', () => {
      component.total = 5;
      component.current = 5;
      component.ngOnInit();
      fixture.detectChanges();

      const ellipsis: DebugElement[] = fixture.debugElement.queryAll(By.css('li.pagination_li.ellipsis'));
      expect(ellipsis.length).toBe(0);

      const renderedText = getRenderedNumbers(fixture);
      expect(renderedText).toContain('1 2 3 4 5');
    });

    it('should correctly render 6 buttons', () => {
      component.total = 6;
      component.current = 5;
      component.ngOnInit();
      fixture.detectChanges();

      const ellipsis: DebugElement[] = fixture.debugElement.queryAll(By.css('li.pagination_li.ellipsis'));
      expect(ellipsis.length).toBe(0);

      const renderedText = getRenderedNumbers(fixture);
      expect(renderedText).toContain('1 2 3 4 5 6');
    });

    it('should correctly render 7 buttons', () => {
      component.total = 7;
      component.current = 4;
      component.ngOnInit();
      fixture.detectChanges();

      const renderedText = getRenderedNumbers(fixture);
      expect(renderedText).toContain('1 2 3 4 5 6 7');
    });
  });

  describe('when dynamically changing inputs', () => {
    let component: PaginationWrapperComponent;
    let fixture: ComponentFixture<PaginationWrapperComponent>;

    beforeEach(async(() => {
      TestBed.configureTestingModule({
        declarations: [ PaginationWrapperComponent, PaginationComponent ],
        imports: [PaginationArrowModule, PaginationNumberModule],
        providers: [ PaginationService ]
      })
      .compileComponents();
    }));

    beforeEach(() => {
      fixture = TestBed.createComponent(PaginationWrapperComponent);
      component = fixture.componentInstance;
      fixture.detectChanges();
    });

    it('should render 7 buttons after changing current page', () => {
      component.currentPage = 1;
      component.totalPages = 7;
      fixture.detectChanges();

      const ellipsis: DebugElement[] = fixture.debugElement.queryAll(By.css('li.pagination_li.ellipsis'));
      expect(ellipsis.length).toBe(1);

      component.currentPage = 4;
      fixture.detectChanges();

      const renderedText = getRenderedNumbers(fixture);
      expect(renderedText).toContain('1 2 3 4 5 6 7');
    });

    it('should disable the next button if total is reduced to current', () => {
      component.currentPage = 5;
      component.totalPages = 10;
      fixture.detectChanges();

      const nextArrow = fixture.debugElement.queryAll(By.directive(PaginationArrowComponent))[1];
      expect(nextArrow.componentInstance.isEnable).toBe(true);

      component.totalPages = 5;
      fixture.detectChanges();

      expect(nextArrow.componentInstance.isEnable).toBe(false);
    });

    it('should set current to total when total decreases below current page value', () => {
      component.currentPage = 5;
      component.totalPages = 5;
      fixture.detectChanges();

      component.totalPages = 3;
      fixture.detectChanges();

      const paginationComponent = fixture.debugElement.query(By.directive(PaginationComponent));
      expect(paginationComponent.componentInstance.paginationService.currentPage).toBe(component.totalPages);
    });

    it('should render correctly after changing group sizes', () => {
      component.currentPage = 3;
      component.totalPages = 5;
      fixture.detectChanges();

      component.firstGroupNum = 1;
      component.midGroupNum = 1;
      component.lastGroupNum = 1;
      fixture.detectChanges();

      const ellipsis: DebugElement[] = fixture.debugElement.queryAll(By.css('li.pagination_li.ellipsis'));
      expect(ellipsis.length).toBe(2);

      const renderedText = getRenderedNumbers(fixture);
      expect(renderedText).toContain('1 3 5');
    });
  });
});

function getRenderedNumbers(fixture) {
  const numberComponents: DebugElement[] = fixture.debugElement.queryAll(By.directive(PaginationNumberComponent));
  const numberComponentsText = numberComponents
    .reduce((accumulator, currentElement) => `${accumulator} ${currentElement.nativeElement.textContent}`, '');
  return numberComponentsText;
}
