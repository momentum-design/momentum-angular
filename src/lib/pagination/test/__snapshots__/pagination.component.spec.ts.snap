// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComponent creation and rendering should create and match snapshot 1`] = `
<md-pagination
  build={[Function Function]}
  current={[Function Number]}
  firstGroupEnd={[Function Number]}
  firstGroupNum={[Function Number]}
  handleKeydown={[Function Function]}
  href=""
  hrefReplaceReg={[Function String]}
  ifPreventDefault="false"
  lastGroupNum={[Function Number]}
  lastGroupStart={[Function Number]}
  maxCurrentForLastGroup={[Function Number]}
  midGroupEnd={[Function Number]}
  midGroupNum={[Function Number]}
  midGroupStart={[Function Number]}
  minCurrentForFirstGroup={[Function Number]}
  paginationService={[Function PaginationService]}
  renderGroup={[Function Array]}
  renderGroupFirst={[Function Array]}
  renderGroupLast={[Function Array]}
  renderGroupMid={[Function Array]}
  total={[Function Number]}
  unsubscribe$={[Function Subject]}
  whenClick={[Function EventEmitter]}
>
  <md-pagination-arrow
    class="pagination-previous pagination_li md-icon icon pagination_disable icon-arrow-left-optical_20"
    ng-reflect-if-prevent-default="false"
    ng-reflect-is-previous="true"
  >
    
    <span
      class="show-for-sr"
    >
      page
    </span>
    
  </md-pagination-arrow><md-pagination-number
    class="pagination_li pagination_current"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="1"
  >
    
    1
    
  </md-pagination-number><md-pagination-number
    class="pagination_li"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="2"
  >
    
    <a
      aria-label="Page 2"
      class="pagination_li_a"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      2
    </a>
    
  </md-pagination-number><md-pagination-number
    class="pagination_li"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="3"
  >
    
    <a
      aria-label="Page 3"
      class="pagination_li_a"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      3
    </a>
    
  </md-pagination-number><md-pagination-number
    class="pagination_li"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="4"
  >
    
    <a
      aria-label="Page 4"
      class="pagination_li_a"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      4
    </a>
    
  </md-pagination-number><li
    aria-hidden="true"
    class="pagination_li ellipsis"
  >
    ...
  </li><md-pagination-number
    class="pagination_li"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="9"
  >
    
    <a
      aria-label="Page 9"
      class="pagination_li_a"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      9
    </a>
    
  </md-pagination-number><md-pagination-number
    class="pagination_li"
    ng-reflect-if-prevent-default="false"
    ng-reflect-index="10"
  >
    
    <a
      aria-label="Page 10"
      class="pagination_li_a"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      10
    </a>
    
  </md-pagination-number><md-pagination-arrow
    class="pagination-previous pagination_li md-icon icon pagination_li_last"
    ng-reflect-if-prevent-default="false"
    ng-reflect-is-previous="false"
  >
    
    <a
      aria-label="Next page"
      class="pagination_li_a icon--b1 md-icon icon icon-arrow-right_20"
      href="javascript:void(0)"
      ng-reflect-ng-class="[object Object]"
    >
      <span
        class="show-for-sr"
      >
        page
      </span>
    </a>
    
  </md-pagination-arrow>
</md-pagination>
`;
